squads/dev-app/LuciaAuth/credentials.md

```diff
@@ -0,0 +1,458 @@
+
+export async function POST(req: Request) {
+	const { email, password } = await req.json();
+
+	if (!isValidEmail(email)) {
+		return Response.json("Email inválido.", { status: 400 });
+	}
+
+	if (password.lenght < 8) {
+		return Response.json("Senha inválida.", { status: 400 });
+	}
+
+	const user = await prisma.user.findUnique({ where: { email: email } });
+
+	if (!user) {
+		const salt = bcrypt.genSaltSync(10);
+		bcrypt.hashSync(password, salt); // Mascarar o tempo de resposta para possíveis atacantes. Não necessário!
+		return Response.json("Email ou senha não válidos", { status: 400 });
+	}
+
+	const hashed_password = user.hashed_password;
+
+	if (!hashed_password) {
+		return Response.json("Unauthorized", { status: 401 });
+	}
+	const validPassword = bcrypt.compareSync(password, hashed_password);
+
+	if (!validPassword) {
+		return Response.json("Email ou senha não válidos", { status: 400 });
+	}
+
+	const session = await lucia.createSession(user.id, {});
+	const sessionCookie = lucia.createSessionCookie(session.id);
+	return Response.json(null, {
+		status: 302,
+		headers: {
+			Location: "/profile",
+			"Set-Cookie": sessionCookie.serialize(),
+		},
+	});
+}
+
+```
+===
+
+## Validar sessão/ Verificar usuário
+
+!!!
+Para fazer validações, o backend precisa verificar se a request de validação ocorreu por meio de um [CSRF.](https://www.treinaweb.com.br/blog/cross-site-request-forgery-csrf-e-abordagens-para-mitiga-lo)
+
+==- Exemplo Next
+Para o next esse ataque pode ser evitado utilizando o middleware.ts
+
+```ts middleware.ts
+import { verifyRequestOrigin } from "lucia";
+import { NextResponse } from "next/server";
+import type { NextRequest } from "next/server";
+
+export async function middleware(request: NextRequest): Promise<NextResponse> {
+	if (request.method === "GET") {
+		return NextResponse.next();
+	}
+	const originHeader = request.headers.get("Origin");
+	const hostHeader = request.headers.get("Host");
+	if (
+		!originHeader ||
+		!hostHeader ||
+		!verifyRequestOrigin(originHeader, [hostHeader])
+	) {
+		return new NextResponse(null, {
+			status: 403,
+		});
+	}
+	return NextResponse.next();
+}
+
+```
+===
+!!!
+
+Como diversas páginas podem necessitar do usuário, uma boa ideia é criar uma função para não precisar escrever tudo em várias páginas. A função abaixo retornará um `user` se ele estiver "logado".
+
+==- Função para pedir usuário
+```ts utils/getUser.ts
+import { cookies } from "next/headers";
+import { cache } from "react";
+
+import type { Session, User } from "lucia";
+import { lucia } from "../../auth/lucia";
+
+export const validateRequest = cache(
+	async (): Promise<
+		{ user: User; session: Session } | { user: null; session: null }
+	> => {
+		const sessionId = cookies().get(lucia.sessionCookieName)?.value ?? null;
+		if (!sessionId) {
+			return {
+				user: null,
+				session: null,
+			};
+		}
+
+		const result = await lucia.validateSession(sessionId);
+		try {
```

Comment:
Acho melhor não usar o `validateRequest` não porque pode acabar confundindo a galera, achando que tem que revalidar a sessão pra invalidá-la.

Reply:
Só faz verificando se há cookies, se houver, invalida a sessão pelo id, independente se a sessão for válida ou não.